antsibull/build_ansible_commands.py

```diff
@@ -19,38 +19,56 @@
 import sh
 from jinja2 import Template
 from packaging.version import Version as PypiVer
+from semantic_version import Version as SemVer
 
 from . import app_context
 from .build_changelog import ReleaseNotes
 from .changelog import get_changelog
 from .collections import install_separately, install_together
 from .dependency_files import BuildFile, DependencyFileData, DepsFile
-from .galaxy import CollectionDownloader
+from .galaxy import CollectionDownloader, DownloadResults
 
 
 #
 # Common code
 #
 
 
-async def download_collections(deps, galaxy_url, download_dir, collection_cache=None):
+async def download_collections(deps: t.Dict[str, str],
+                               galaxy_url: str,
+                               download_dir: str,
+                               collection_cache: t.Optional[str] = None,
+                               force_versions: t.Optional[t.Mapping[str, SemVer]] = None
+                               ) -> t.Dict[str, SemVer]:
     requestors = {}
     async with aiohttp.ClientSession() as aio_session:
         lib_ctx = app_context.lib_ctx.get()
         async with asyncio_pool.AioPool(size=lib_ctx.thread_max) as pool:
             downloader = CollectionDownloader(aio_session, download_dir,
                                               collection_cache=collection_cache,
                                               galaxy_server=galaxy_url)
+
+            async def download_specific(collection_name_: str, version: SemVer):
+                download_path = await downloader.download(collection_name_, version)
+                return DownloadResults(version=version, download_path=download_path)
+
             for collection_name, version_spec in deps.items():
-                requestors[collection_name] = await pool.spawn(
-                    downloader.download_latest_matching(collection_name, version_spec))
+                force_version = (
+                    None if force_versions is None else force_versions.get(collection_name)
+                )
+                if force_version is None:
+                    requestors[collection_name] = await pool.spawn(
+                        downloader.download_latest_matching(collection_name, version_spec))
+                else:
+                    requestors[collection_name] = await pool.spawn(
+                        download_specific(collection_name, force_version))
```

Comment:
Reading the calling code for this function, I think we can simplify the logic by using empty dicts instead of None.  We still need None in case the keyword arg is not supplied but that can be processed at the beginning of this function into an empty dictionary.

Reply:
Implemented in 04ea5970bd00d7ba4485343bd43b823692eda1de.